### DISCUSSION TITLE
Add idx to router.object

### DISCUSSION THREAD

**Discussion by tatethurston** [vercel/next.js#34980] [Ideas]
### Describe the feature you'd like to request

I'd like to propose adding [idx](https://github.com/vercel/next.js/blob/a51803624638a626694b03c23b48e373eb586517/packages/next/shared/lib/router/router.ts#L1410) to the [router.object](https://nextjs.org/docs/api-reference/next/router#router-object).`idx` is included in the history state on every new history entry.

This is valuable for reporting purposes such as the number of page views in the current session. 

It's also valuable when building native-like applications using nextjs, particularly for back button behaviors. When replicating a native like experience, it's helpful to be able to check if a programatic `router.back()` would kick the user out of the app, and if so, fallback to another page (like a Homepage). This pattern frequently comes up for back buttons rendered on an error page and settings pages.

### Describe the solution you'd like

A new `idx` field (or perhaps `pageCount` / `pageIndex`?) added to the router object.

### Describe alternatives you've considered

1. Accessing `history.state.idx` directly. This works, but is undocumented and I'm concerned about the field changing or being removed in a next update and application code silently breaking.

2. Another option would be to include a `fallback` option with [router.back](https://nextjs.org/docs/api-reference/next/router#routerback) so that users aren't backed out of the application.

**Comment by AkifumiSato** [comment]
+1
Created an implementation PR to add idx to router.
#35681

**Reply by mramadanahmed** [reply]
@AkifumiSato Thanks for the idx feature it is very useful one, but can it be persisted in the session storage? as of now if the user refresh the page it will be rested to 0

#### Case Study
A user open the login page and enter his email, click continue and then he redirected to the otp page. If he decided to return back to the login screen we need his email to be persist. He can use  browser back or button with router.back(). 

#### Implementation
The idx was a saver for me.  I used it to build a full navigation history array in the session storage. The idea behind that is to keep some state for the current page based on the uid of the page in the history stack. So once the user navigate back to it, I can show the page with its state where he left it. 

I had a useEffect in the _app to listen to the next router routeChangeComplete to maintain an array of the visited pages with its state if any in the session storage. Using that I was able to push/pop items from my state array based on the idx value as unique key.

Saving that state based only on the url the value and not based on the idx, the persist value will be shown to the user whenever he navigate to login page until we clear it based on some event like a clear button. 

The problem is once the user refresh the page the idx counter reset to 0 so the whole structure is collapsed.

**Reply by tatethurston** [reply]
I think this ask should be a separate feature request.

**Reply by mramadanahmed** [reply]
@tatethurston Many thanks for your quick response. Sure it could be a new feature request. However, I posted that here because I saw it as a bug with the second point in your request regarding to the back fallback

> Another option would be to include a fallback option with [router.back](https://nextjs.org/docs/api-reference/next/router#routerback) so that users aren't backed out of the application.

I achieved the same as well using the idx, but lets take the following scenario. 

The user open the website and  navigate until the idx become 4, then eventually the user refresh the page from the address bar. Now the idx will be 0 not 4 and user will be stuck in the fallback in that page despite he has 4 steps back still valid inside the application

Hopefully if you already implement the back button fallback to confirm the behavior.

Kindly recommend

**Reply by tatethurston** [reply]
@mramadanahmed You are right — great catch. That would be a useful property for the back button behavior.

**Reply by mramadanahmed** [reply]
Bug #36662 is created to track that issue

**Reply by AkifumiSato** [reply]
I have already created a pull request, as I put on the issue.
I don't know if it will be merged or not since there is no response.
Any help would be greatly appreciated!
